### ISSUE TITLE
RpcReply.OK not set correctly due to XML unmarshalling of boolean is broken

### ISSUE THREAD

**Issue by emasean** [Juniper/go-netconf#34]
The following rpc-reply is received:

     `<rpc-reply message-id="101"
                xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
       <ok/>
     </rpc-reply>`

But the unmarshalled `RpcReply.ok` did not set to `true`.

After checking the unmarshalling format:

`// RPCReply defines a reply to a RPC request
type RPCReply struct {
	XMLName  xml.Name   `xml:"rpc-reply"`
	Errors   []RPCError `xml:"rpc-error,omitempty"`
	Data     string     `xml:",innerxml"`
	Ok       bool       `xml:",omitempty"`
	RawReply string     `xml:"-"`
}`

the boolean unmarshalling seems to be broken: accprdomg to the following[site](https://stackoverflow.com/questions/23724591/golang-unmarshal-self-closing-tags)

So making the following RPCReply:
`type RPCReply struct {
	XMLName  xml.Name   `xml:"rpc-reply"`
	Errors   []RPCError `xml:"rpc-error,omitempty"`
	Data     string     `xml:",innerxml"`
	Ok       *struct{}       `xml:"ok,omitempty"`
	RawReply string     `xml:"-"`
}`

and check if `ok != nil` will make it work

**Comment by nemith** [comment]
I think we can create a custom unmarshaler here.

**Comment by wtucker** [comment]
Does the RPCReply.Ok flag tell us anything useful?  From what I've been able to find the RPCReply is only returned if no error was generated.  If something does go wrong, RPCError is raised instead.

**Comment by wtucker** [comment]
Sample of `validate` response with bad config (should raise RPCError):

```
<rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:junos="http://xml.juniper.net/junos/15.1F4/junos">
<commit-results>
<rpc-error>
<error-type>application</error-type>
<error-tag>invalid-value</error-tag>
<error-severity>error</error-severity>
<error-path>[edit]</error-path>
<error-message>mgd: Missing mandatory statement: 'root-authentication'</error-message>
<error-info>
<bad-element>system</bad-element>
</error-info>
</rpc-error>
<rpc-error>
<error-type>protocol</error-type>
<error-tag>operation-failed</error-tag>
<error-severity>error</error-severity>
<error-message>
configuration check-out failed: (missing mandatory statements)
</error-message>
</rpc-error>
</commit-results>
</rpc-reply>
```

**Comment by wtucker** [comment]
Sample from a lab QFX that had MPLS enabled but not licensed:

```
<rpc><validate><source><candidate/></source></validate></rpc>
<rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:junos="http://xml.juniper.net/junos/16.1R3/junos">
<commit-results>
<rpc-error>
<error-severity>warning</error-severity>
<error-path>[edit protocols]</error-path>
<error-message>mgd: requires 'mpls' license</error-message>
<error-info>
<bad-element>mpls</bad-element>
</error-info>
</rpc-error>
<rpc-error>
<error-severity>warning</error-severity>
<error-path>[edit protocols]</error-path>
<error-message>mgd: requires 'bgp' license</error-message>
<error-info>
<bad-element>bgp</bad-element>
</error-info>
</rpc-error>
<routing-engine junos:style="normal">
<name>fpc0</name>
<commit-check-success/>
</routing-engine>
</commit-results>
<ok/>
</rpc-reply>
]]>]]>
```

**Comment by yhzs8** [comment]
> Does the RPCReply.Ok flag tell us anything useful? From what I've been able to find the RPCReply is only returned if no error was generated. If something does go wrong, RPCError is raised instead.

For me it is more intuitive to check if the response is OK first (the normal flow) rather than if response contains error first (the exception flow). I believe this is another workaround that works but The protocol wrapper should work as stated in the RFC: https://tools.ietf.org/html/rfc6241#page-19

**Comment by wtucker** [comment]
> For me it is more intuitive to check if the response is OK first (the normal flow) rather than if response contains error first (the exception flow). I believe this is another workaround that works but The protocol wrapper should work as stated in the RFC: https://tools.ietf.org/html/rfc6241#page-19

I've filed #55 to update documentation so it explains that the Ok flag doesn't mean what one might expect.

There are two branches that fix setting the RPCReply.Ok flag processing, but both have drawbacks.  I'm open to alternate suggestions for handle the unmarshaling.

**Comment by nemith** [comment]
Agreed that it should follow the RFC.  We worked on this package through the NANOG hackathon and may have some ideas on unmarshaling it, but still experimenting.

**Comment by nemith** [comment]
This should work: https://play.golang.org/p/Qyl_i0Ihf96   @wtucker  can you update your code/tests with this method?

**Comment by vithubati** [comment]
why this issue hasn't been fixed yet? I would like to know the updates

**Comment by nemith** [comment]
This is fixed in v2. (I understand the xml package so much better than 10 years ago)
